### PR TITLE
remove left index in async mode

### PR DESCRIPTION
**Pull request by zhoney**
implemented: #247

Change-Id: Ie1baacb93f18d66ed7169161f15f398701e955a3

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 1]
```diff
@@ -124,37 +143,34 @@ private void processRangeIndexLeft(ConditionQuery query,
...
         }
-
         // Delete unused index
         for (IndexLabel il : relatedIndexLabels(deletion)) {
-            if (!CollectionUtil.hasIntersection(il.indexFields(), incorrectPKs)) {
+            Set<Id> incorrectPKIds = incorrectPKs.keySet().stream()
+                                                 .map(PropertyKey::id)
+                                                 .collect(Collectors.toSet());
+            @SuppressWarnings("unchecked")
+            Collection<Id> pks = CollectionUtils.intersection(il.indexFields(),
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: use CollectionUtil.intersect()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 2]
```diff
@@ -124,37 +143,34 @@ private void processRangeIndexLeft(ConditionQuery query,
...
+        HugeElement deletion = this.constructErrorElem(query, element,
+                                                       incorrectPKs);
+        if (deletion == null) {
+            return;
         }
-
         // Delete unused index
         for (IndexLabel il : relatedIndexLabels(deletion)) {
-            if (!CollectionUtil.hasIntersection(il.indexFields(), incorrectPKs)) {
+            Set<Id> incorrectPKIds = incorrectPKs.keySet().stream()
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: rename to incorrectPkIds

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 3]
```diff
@@ -182,6 +198,11 @@ private void processSecondaryOrSearchIndexLeft(ConditionQuery query,
                  */
                 this.updateIndex(il.id(), element, false);
             }
+            this.commit();
+            if (this.deletedByError(element, pks, incorrectPKs)) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: rename pks to incorrectIndexFields

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/EphemeralJobBuilder.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+public class EphemeralJobBuilder<T> {
+
+    private final HugeGraph graph;
+
+    private String name;
+    private String input;
+    private EphemeralJob<T> job;
+
+    private static int tempTaskId = 0;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/EphemeralJobBuilder.java]: unused

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 5]
```diff
@@ -79,6 +86,18 @@ public GraphIndexTransaction(HugeGraph graph, BackendStore store) {
...
 
+    protected Id asyncRemoveIndexLeft(ConditionQuery query,
+                                      HugeElement element) {
+        RemoveLeftIndexCallable callable = new RemoveLeftIndexCallable(query,
+                                                                       element);
+        EphemeralJobBuilder<Object> builder =
+                                    EphemeralJobBuilder.of(this.graph())
+                                                       .name(element.name())
+                                                       .job(callable);
+        HugeTask<?> task = builder.schedule();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]:
```
HugeTask<?> task = EphemeralJobBuilder.of(this.graph())
                                      .name(element.name())
                                      .job(callable)
                                      .schedule();
```

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 6]
```diff
@@ -79,6 +86,18 @@ public GraphIndexTransaction(HugeGraph graph, BackendStore store) {
...
+        return task.id();
+    }
+
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: prefer to move methods associated with removeIndexLeft to class RemoveLeftIndexCallable

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 7]
```diff
@@ -1090,10 +1174,39 @@ public static IndexQueries of(IndexLabel il, ConditionQuery query) {
...
 
-    public static enum OptimizedType {
+    public enum OptimizedType {
         NONE,
         PRIMARY_KEY,
         SORT_KEY,
         INDEX
     }
+
+    public static class RemoveLeftIndexCallable extends EphemeralJob<Object> {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: RemoveLeftIndexJob

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/EphemeralJobBuilder.java (hunk 4 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/job/EphemeralJobBuilder.java]: used for ephemeral task id

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/EphemeralJobBuilder.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        TaskScheduler scheduler = this.graph.taskScheduler();
+        scheduler.schedule(task);
+
+        return task;
+    }
+
+    private Id genTaskId() {
+        if (ephemeralTaskId > 0) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/EphemeralJobBuilder.java]: ephemeralTaskId >= 0

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 9]
```diff
@@ -79,110 +87,15 @@ public GraphIndexTransaction(HugeGraph graph, BackendStore store) {
...
-                 * is correct, the correct prefix part will be deleted,
-                 * so rebuild the index again with the correct property.
-                 */
-                this.updateIndex(il.id(), element, false);
-            }
-        }
+    protected Id asyncRemoveIndexLeft(ConditionQuery query,
+                                      HugeElement element) {
+        RemoveLeftIndexJob callable = new RemoveLeftIndexJob(query,
+                                                                       element);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: align

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 10]
```diff
@@ -1091,10 +1004,220 @@ public static IndexQueries of(IndexLabel il, ConditionQuery query) {
...
+        @Override
+        public String type() {
+            return REMOVE_LEFT_INDEX;
+        }
+
+        @Override
+        public Object execute() {
+            this.tx = this.graph().graphTransaction().indexTransaction();
+            this.removeIndexLeft(this.query, this.element);
+            return null;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: prefer to return removed count

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 11]
```diff
@@ -1091,10 +1003,228 @@ public static IndexQueries of(IndexLabel il, ConditionQuery query) {
...
+        protected int removeIndexLeft(ConditionQuery query,
+                                      HugeElement element) {
+            if (element.type() != HugeType.VERTEX &&
+                element.type() != HugeType.EDGE_OUT &&
+                element.type() != HugeType.EDGE_IN) {
+                throw new HugeException("Only accept element of type VERTEX " +
+                                        "and EDGE to remove left index, " +
+                                        "but got: '%s'", element.type());
+            }
+            int rCount = 0;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: change to long

**Pull request merged**